Simplify exception handling for polyglot exceptions

hubertp opened the pull request:
### Pull Request Description

This removes the special handling of polyglot exceptions and allows matching on Java exceptions in the same way as for any other types.

`Polyglot_Error`, `Panic.catch_java` and `Panic.catch_primitive` are gone

The change mostly deals with the backslash of removing `Polyglot_Error` and two `Panic` methods. 
`Panic.catch` was implemented as a builtin instead of delegating to `Panic.catch_primitive` builtin that is now gone.

This fixes https://www.pivotaltracker.com/story/show/182844611

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
Remove Polyglot_Error and Panic.catch_java

This removes the special handling of polyglot exceptions and allows
matching on Java exceptions in the same way as for any other types.

The change mostly deals with the backslash of removing Polyglot_Error.
Only base stdlib pass so far.

hubertp committed:
Fix remaining issues

hubertp committed:
Add missing TruffleBoundary

hubertp committed:
nits

hubertp committed:
Remove `Panic.catch_primitive`

This was requested to simplify the implementation.

Minor nit - for test purposes, still added it in micro-distribution. If
not necessary, we might need to drop some tests instead.

hubertp committed:
Update changelog

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:
```diff
@@ -167,5 +167,5 @@ type Array
         if Meta.is_same_object Array self then Meta.is_same_object Array that else
             eq_at i = self.at i == that.at i
-            Panic.catch_primitive handler=(_ -> False)
+            Panic.catch Any handler=(_ -> False)
```
No more `catch_primitive`. That's great!

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Json.enso:
```diff
@@ -53,8 +53,8 @@ type Json
              example_parse = Json.parse '{ "a": 1 }'
     parse : Text -> Json ! Json_Parse_Error
     parse json_text =
-        Panic.catch_java Any (Internal.parse_helper json_text) java_exception->
```
No more `catch_java`. That's is perfect!

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/IsPayloadOfPanicTypeNode.java:
```diff
@@ -0,0 +1,70 @@
+/** An implementation of the payload check against the expected panic type. */
+@NodeInfo(shortName = "IsPayloadOfPanicTypeNode")
+public abstract class IsPayloadOfPanicTypeNode extends Node {
+  private @Child IsSameObjectNode isSameObject = IsSameObjectNode.build();
```
Should this logic be the same as in [Meta.is_a](https://github.com/enso-org/enso/pull/3949) check and `case_of_on_type` check? Shouldn't we have a single `IsATypeNode` and share it among all the implementations?

hubertp replied:
It's a subset, I think. Since I don't care about things such as strings, arrays, numbers, dates etc

radeusgd replied:
I agree with @JaroslavTulach that it seems best to have a single implementation for these.

@hubertp why do you think it is a subset? Does the `IsPayloadOfPanicTypeNode` not check for strings and numbers?

Will `Panic.catch Integer (Panic.throw 42)` work?  I'm not certain that it has to, but it used to work.

Surely, as long as `Panic.throw` will accept a type of payload, the same type of payload must be catchable by `Panic.catch`.

So if we want to disallow catching integers, we need to disallow throwing them as well IMO.

hubertp replied:
Yes, I knew that there will be some intersection with `Meta.is_a` builtin work by @JaroslavTulach but I didn't want this to be a blocker.
And yes, the limitation was on purpose. But it seems that `Panic.throw` doesn't make that distinction so I should relax indeed (feels wrong). 

So should I wait on `Meta.is_a` work to get in first? @JaroslavTulach any ETA when will this get in?

radeusgd replied:
Feels wrong to catch an Integer? I guess we could disallow throwing it too. But not sure where to draw the line? Just disallow primitive types (Text, Number, maybe Array)?

hubertp replied:
> Feels wrong to catch an Integer? I guess we could disallow throwing it too. But not sure where to draw the line? Just disallow primitive types (Text, Number, maybe Array)?

Similarly to Java/Scala I would expect the value to be thrown to be of a known base type (there `Throwable`). It does not appear like we use this in our tests but surprisingly we have an example with that in `Panic.throw`. We should certainly have tests for things that we have in the _example_ section.

hubertp replied:
I've been looking at this piece of code and actually I was wrong about 

> It's a subset, I think. Since I don't care about things such as strings, arrays, numbers, dates etc

Since the code is very similar to `CatchTypeBranchNode` it is only polyglot arrays.
Also looking at the code again, I discovered another bug in the existing code of
```
case Integer.valueOf "1" of
    _ : Any -> 0
    _ -> 1
```
would return `1` rather than `0`. I fixed that in `catch` but not in the `case` branch check.

So I will try to handle polyglot arrays (since I guess one could throw them, for whatever reason) and port a fix for the above. But I would delay merging with `Meta.is_a` when it actually lands.

hubertp replied:
So `CatchTypeBranchNode` is now merged with `IsValueOfTypeNode` so one less thing to deal with.

JaroslavTulach replied:
> the value to be thrown to be of a known base type (there Throwable)

There is also the _"JavaScript way"_ - you can throw anything - strings, numbers, etc.

JaroslavTulach replied:
> So `CatchTypeBranchNode` is now merged with `IsValueOfTypeNode` so one less thing to deal with.

OK, so since 61d23a44e0a1e4ead6ad79544506be28d35760b9 the _type checking_ code shall live in `IsValueOfTypeNode` and be shared. I'll count with that in my #3949.

radeusgd replied:
> > the value to be thrown to be of a known base type (there Throwable)
> 
> There is also the _"JavaScript way"_ - you can throw anything - strings, numbers, etc.

Very good point!

```
from Standard.Base import all

foreign js foo = """
    return 42+23;

foreign js bar = """
    throw 42;

main =
    IO.println foo
    r = Panic.catch Any bar err->
        IO.println err
        "error handled"
    IO.println r
```

```
65
(Panic (Polyglot_Error.Error 42) 42)
error handled
```

Our current `Panic.catch` is capable of catching JS exceptions, so probably would be good to keep that and ideally would be great if I could also do `Panic.catch Integer` there - currently it does not work due to `Polyglot_Error` wrapper - I did not check if on your branch it now works.

Either way - if possible it would be great to add a test case to `Error_Spec` checking catching the JS exception - and if possible also checking there if filtering like `Panic.catch Integer js_function_throwing_js_int` works.

hubertp replied:
> Either way - if possible it would be great to add a test case to `Error_Spec` checking catching the JS exception - and if possible also checking there if filtering like `Panic.catch Integer js_function_throwing_js_int` works.

This proved to be a bit more painful than I initially imagined but that is now possible (as checked by additional tests).

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Error.java:
```diff
@@ -58,7 +57,6 @@ public Error(Builtins builtins, EnsoContext context) {
     uninitializedState = builtins.getBuiltinType(UninitializedState.class);
     noSuchMethod = builtins.getBuiltinType(NoSuchMethod.class);
     noSuchConversion = builtins.getBuiltinType(NoSuchConversion.class);
-    polyglotError = builtins.getBuiltinType(PolyglotError.class);
```
Polyglot error is gone. Great.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java:
```diff
@@ -11,6 +11,7 @@
 import com.oracle.truffle.api.source.SourceSection;
 import org.enso.interpreter.node.expression.builtin.text.util.TypeToDisplayTextNode;
 import org.enso.interpreter.node.expression.builtin.text.util.TypeToDisplayTextNodeGen;
+import org.enso.interpreter.runtime.callable.atom.Atom;
```
Probably unused import.

JaroslavTulach commented on engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/CompileDiagnosticsTest.scala:
```diff
@@ -12,9 +12,10 @@ class CompileDiagnosticsTest extends InterpreterTest {
           |
           |main =
-          |    x = Panic.catch_primitive () .convert_to_dataflow_error
+          |    x = Panic.catch Any () .convert_to_dataflow_error
```
Using `catch Any` is even six characters shorter than `catch_primitive`!

JaroslavTulach commented on engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/SystemProcessTest.scala:
```diff
@@ -55,7 +55,7 @@ class SystemProcessTest extends InterpreterTest with OsSpec {
 
       val error = the[InterpreterException] thrownBy eval(code)
-      error.getMessage should include("a polyglot object")
+      error.getMessage should include("java.io.IOException")
```
Yup, that's more valuable.

JaroslavTulach reviewed (approved):
Great step forward. The only thing I'd like to do at one point is the share a single `IsATypeNode` check implementation among all _type checking_ code in the engine.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Network/URI.enso:
```diff
@@ -39,8 +39,8 @@ type URI
-        Panic.catch_java Any (URI.Value (Java_URI.create text)) java_exception->
-            Error.throw (Syntax_Error.Error ("URI syntax error: " + java_exception.getMessage))
+        Panic.catch Any (URI.Value (Java_URI.create text)) caught_panic->
+            Error.throw (Syntax_Error.Error ("URI syntax error: " + caught_panic.payload.getMessage))
```
Shouldn't this catch a `java.lang.Exception`? We change semantics here.

radeusgd commented on distribution/lib/Standard/Image/0.0.0-dev/src/Data/Image.enso:
```diff
@@ -69,7 +69,7 @@ type Image
                 if flags.is_empty then Java_Codecs.READ_FLAG_EMPTY else
                     flags.map .to_integer . reduce (_.bit_or _)
             _ -> flags.to_integer
-        Panic.catch_java Any (Image.Value (Java_Codecs.read path read_flags)) _->
+        Panic.catch Any (Image.Value (Java_Codecs.read path read_flags)) _->
             Error.throw (File_Error.IO_Error (File.new path) 'Failed to read the file')
```
Again, we change semantics from matching _any **Java**_ exception to matching _any_ panic whatsoever. This is an issue, because if the user provides wrong types of arguments to `path` or `read_flags` the any `Type_Error` or `Unsupported_Argument_Types` will enter the same path as Java exceptions and result in an `IO_Error` - which is wrong.

Whenever we had `catch_java Any` we need something like `catch Java_Exception` to preserve the original semantics. The `catch_java Any` in such places was used _on purpose_ instead of `catch Any`, to correctly distinguish errors from within the polyglot helper from errors related to the helper not being fired at all (mostly due to type mismatches).

hubertp replied:
>  The catch_java Any in such places was used on purpose instead of catch Any

OK, I will make the necessary changes.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/CatchPanicNode.java:
```diff
@@ -54,24 +61,28 @@ Object doExecute(
     }
   }
 
   private Object executeCallback(
```
I'd rename this method.

It now can _either_ execute a callback _or_ forward the panic further.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/CatchPanicNode.java:
```diff
@@ -54,24 +61,28 @@ Object doExecute(
+    if (profile.profile(isPayloadOfPanicTypeNode.execute(panicType, payload))) {
+      return invokeCallableNode.execute(handler, frame, state, new Object[] {caughtPanic});
+    } else {
+      return throwPanicNode.execute(caughtPanic);
```
Hmm, do we need the `throwPanicNode` here?

I'd think in this branch there should be no need to even allocate the `caughtPanic` Atom and we should be able to just `throw interopLibrary.throwException(originalException)`. It should be faster too.

hubertp replied:
Adding a short-circuit that avoids the additional node will work as well. Was trying to avoid DRY.

> It should be faster too.

Whenever exceptions start to float around, one should generally not expect some peak performance.

radeusgd reviewed (commented):
Please take care to not change semantics when moving from `catch_java Any`.

hubertp committed:
nit

hubertp committed:
PR review

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/text/util/TypeToDisplayTextNode.java:
```diff
@@ -71,6 +72,12 @@ String doDisplay(
       } catch (UnsupportedMessageException e) {
         throw new IllegalStateException("Receiver declares a meta object, but does not return it.");
       }
+    } else if (exceptions.hasExceptionMessage(value)) {
+      try {
+        return strings.asString(exceptions.getExceptionMessage(value));
+      } catch (UnsupportedMessageException e) {
+        throw new IllegalStateException(e);
+      }
```
This looks unexpected.

When we try to get a name of the exception _type_ we return its _message_?? I don't get it. Why?

hubertp replied:
Weird, isn't it? 
Well, there is a reason for it:
- https://github.com/enso-org/enso/blob/develop/test/Tests/src/Semantic/R_Interop_Spec.enso#L144
- https://github.com/enso-org/enso/blob/develop/test/Tests/src/Semantic/R_Interop_Spec.enso#L154

That's needed to keep it that way. `RErrors` are wrapped in `PolyglotExceptionProxy`. The question is, are those tests broken in the first place?

radeusgd replied:
Interesting!

But wait - how does `error.to_text` invoke `TypeToDisplayTextNode` instead of something along the lines of `AnyToDisplayTextNode`?

Oh now I see `AnyToDisplayTextNode` just delegates to `TypeToDisplayTextNode`?? _This_ is weird IMO. Shouldn't we have the exception special handling _there_ instead?

radeusgd replied:
Seems to me that `TypeToDisplayTextNode` is a wrong place for this special case. Can we have it in `AnyToDisplayTextNode` or somewhere similar?

hubertp replied:
Resolved

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java:
```diff
@@ -38,23 +38,23 @@ public EnsoFile(TruffleFile truffleFile) {
 
   @Builtin.Method
-  @Builtin.WrapException(from = IOException.class, to = PolyglotError.class)
+  @Builtin.WrapException(from = IOException.class, to = PanicException.class)
```
Can we make `PanicException` a default here?

radeusgd commented on lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/builtins/SafeWrapException.java:
```diff
@@ -24,21 +25,29 @@ public record SafeWrapException(Attribute.Class from, Attribute.Class to) {
                 "    throw new PanicException(e.getMessage(), this);"
             );
         } else {
-            return List.of(
+            if (to.equals(PanicExceptionClassName)){
```
totally nitpick but IMHO
```
} else if (...) {
    ...
} else {
    ...
}
```
looks better than
```
} else {
    if (...) {
        ...
    } else {
        ...
    }
}
```

radeusgd commented on test/Tests/src/Data/Time/Date_Spec.enso:
```diff
@@ -430,20 +429,20 @@ js_set_zone local_datetime =
 
 js_date year month=1 day=1 =
-    Panic.catch Any (js_set_zone (js_date_impl year month day)) (err -> Error.throw (Time_Error.Error err.payload.cause))
+    Panic.catch Any (js_set_zone (js_date_impl year month day)) (err -> Error.throw (Time_Error.Error err.payload))
```
Really glad of all the removed `.cause`. They were mostly noise.

radeusgd commented on test/micro-distribution/lib/Standard/Base/0.0.0-dev/src/Panic.enso:
```diff
@@ -1,4 +1,7 @@
     throw payload = @Builtin_Method "Panic.throw"
-    catch_primitive ~action handler = @Builtin_Method "Panic.catch_primitive"
+    catch panic_type ~action handler = @Builtin_Method "Panic.catch"
+    catch_primitive ~action handler = Panic.catch Any action handler
```
Why do we still define `catch_primitive` here? Wasn't it removed from the normal distribution? Suspicious.

hubertp replied:
Yes, it is removed from the distribution.
We had a few interpreter tests that used to do
```
Panic.catch_primitive @ caught-> caught.payload
```
which return a syntax error. Except, when translating to `Panic.catch Any @ caught-> caught.payload` they would fail to catch it. 

In reality the application is broken because of a broken IR.
It's a bug in the parser apparently (https://www.pivotaltracker.com/story/show/184054024) and I've been scratching my head around it for a while. In the interest of moving on I added this workaround, which can be obviously removed once the fix is in place.

radeusgd reviewed (approved):
Looks good to me overall, but I'd like to resolve [the discussion on consistency of `Panic.catch` with `Meta.is_a` and `Panic.throw`](https://github.com/enso-org/enso/pull/3981/files#r1050673789) before we proceed.

hubertp committed:
Simplify to_x_text for PolyglotExceptionProxy

hubertp committed:
Catch polyglot arrays with Array type

Also added a case to deal with case pattern matching on Any type that
should match against any value.

hubertp committed:
PR review

hubertp committed:
Merge type match node with is panic of type node

The two were pretty much identical so it makes sense to DRY.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/CatchTypeBranchNode.java:
```diff
@@ -39,48 +28,12 @@ public abstract class CatchTypeBranchNode extends BranchNode {
-        tpe = superTpe;
-        superTpe = superTpe.getSupertype();
-      }
+  public void execute(VirtualFrame frame, Object state, Object value) {
```
Great simplification. I am thrilled to see this PR merged!

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsValueOfTypeNode.java:
```diff
@@ -0,0 +1,96 @@
+
+  @Specialization
+  boolean doAtom(AtomConstructor consr, Object payload) {
+    Object panicType = consr.getType();
```
Is this correct? You ignore the `AtomConstructor` and take it's type. Would that mean that in
```
t v = case v of
   x : Cons -> "Cons"
   x : Nil -> "Nil"
   _ -> "Anything"
```
we would get `t Cons` to be `"Cons"`, `t Nil` to be also `"Cons"`? Shouldn't this return `"Anything"` as none of the parameters is actually a `Type`? At least that's the logic I am trying to implement in

hubertp replied:
Constructors aren't allowed in type patterns:
```
The name List.Cons resolved to a constructor, but constructors are not allowed in type pattern case.
```

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsValueOfTypeNode.java:
```diff
@@ -0,0 +1,96 @@
+    return execute(panicType, payload);
+  }
+
+  @Specialization(guards = {"interop.isMetaObject(panicType)"})
```
I am surprised to not see specialization for `Type`. There is [one in is_a PR](https://github.com/enso-org/enso/pull/3949/files#r1051858704).

hubertp replied:
That reshuffle revealed also a bug in one of our tests when
```
case Array of
    Array_Module -> 1
    _ : Array -> 2
    Array -> 3
    _ -> 4
```
returned `2` instead of `3`. Due to the way we have to deal with arrays this wasn't a straightforward change but we now specialize on `Type`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/text/util/TypeToDisplayTextNode.java:
```diff
@@ -31,6 +31,7 @@ String doDisplay(
       Object value,
       @CachedLibrary(limit = "5") InteropLibrary objects,
       @CachedLibrary(limit = "5") InteropLibrary displays,
+      @CachedLibrary(limit = "5") InteropLibrary exceptions,
```
Probably unused, right?

JaroslavTulach commented on lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/builtins/SafeWrapException.java:
```diff
@@ -13,34 +13,46 @@
  * extracted
  */
-public record SafeWrapException(Attribute.Class from, Attribute.Class to) {
+public record SafeWrapException(Attribute.Class from, Optional<Attribute.Class> to) {
```
Ah, `Optional`... `null` is the Java way!

hubertp replied:
I'm not a fan of NPEs, clearly :)

hubertp committed:
Reshuffle specializations

Specialize on Type in `IsValueOfTypeNode` which broke a number of
assumptions.
Added specialization on `PolyglotExceptionProxy` so that we can match on
the exception thrown directly in JS, as requested by @radeusgd.

Discovered a bug in our tests where we were matching Array type when
type pattern matching.

hubertp commented on test/Tests/src/Semantic/Case_Spec.enso:
```diff
@@ -85,15 +85,22 @@ spec = Test.group "Pattern Matches" <|
             Array -> Test.fail "Expected an array type to match."
             _ : Array -> Nothing
-            _ -> Test.fail "Expected an array type  to match."
-        case Array of
```
This was an invalid test.

hubertp committed:
Added a comment to rememember about a workaround

hubertp committed:
Preventing potential SOs
